staking/staking_hub.go

```diff
@@ -0,0 +1,416 @@
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/0xPolygon/minimal/types"
+	"github.com/hashicorp/go-hclog"
+)
+
+type StakingEventType string
+
+var (
+	StakingEvent   StakingEventType = "staking"
+	UnstakingEvent StakingEventType = "unstaking"
+	UnknownEvent   StakingEventType = "unknown"
+)
+
+var (
+	defaultFileName = "staking-map.json"
+)
+
+// StakingHub acts as a hub (manager) for staked account balances
+type StakingHub struct {
+	// Address -> Stake
+	StakingMap map[types.Address]*big.Int
+
+	// Dirty staking map that's modified during block execution
+	// Address -> Stake
+	DirtyStakingMap map[types.Address]*big.Int
+
+	// Specifies the working directory for the Polygon SDK (location for writeback)
+	WorkingDirectory string
+
+	// Write-back period (in s) for backup staking data
+	WritebackPeriod time.Duration
+
+	// Logger for logging errors and information
+	Logger hclog.Logger
+
+	// Event list mutex
+	EventQueueMutex sync.Mutex
+
+	// Staking map mutex
+	StakingMutex sync.Mutex
+
+	// Dirty staking map mutex
+	DirtyStakingMutex sync.Mutex
+
+	// Close channel
+	CloseCh chan struct {
+	}
+}
+
+var stakingHubInstance StakingHub
+var once sync.Once
+
+// GetStakingHub initializes the stakingHubInstance singleton
+func GetStakingHub() *StakingHub {
+	once.Do(func() {
+		stakingHubInstance = StakingHub{
+			StakingMap:      make(map[types.Address]*big.Int),
+			DirtyStakingMap: make(map[types.Address]*big.Int),
+			CloseCh:         make(chan struct{}),
+			WritebackPeriod: 60 * time.Second,
+		}
+	})
+
+	return &stakingHubInstance
+}
+
+// SetWorkingDirectory sets the writeback directory for the staking map
+func (sh *StakingHub) SetWorkingDirectory(directory string) {
+	sh.WorkingDirectory = directory
+
+	err := sh.readFromDisk()
+	if err != nil {
+		// Log as warning because reading this map is not crucial
+		sh.log(err.Error(), logWarning)
+	}
+
+	go sh.saveToDisk()
+}
+
+// LogType defines possible log types for the logger
+type LogType string
+
+var (
+	logInfo    LogType = "info"
+	logError   LogType = "error"
+	logWarning LogType = "warning"
+)
+
+// log logs the output to the console if the logger is set
+func (sh *StakingHub) log(output string, logType LogType) {
+	if sh.Logger != nil {
+		switch logType {
+		case logInfo:
+			sh.Logger.Info(output)
+		case logError:
+			sh.Logger.Error(output)
+		case logWarning:
+			sh.Logger.Warn(output)
+		}
+	}
+}
+
+// SetLogger sets the StakingHub logger
+func (sh *StakingHub) SetLogger(logger hclog.Logger) {
+	sh.Logger = logger.Named("staking-hub")
+}
+
+// CloseStakingHub stops the writeback process
+func (sh *StakingHub) CloseStakingHub() {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	// Alert the closing channel
+	select {
+	case sh.CloseCh <- struct{}{}:
+		sh.log("Closing staking hub...", logInfo)
+	default:
+		sh.log("SH Writer not set up. Closing staking hub...", logInfo)
+	}
```

Comment:
I didn't understand what you want to do here 100%, but there is a possibility that nothing is sent to channel if you add default case. To avoid this, you can set channel capacity to 1 or remove default case.